[MM-52303] Desktop development docs update (#1248)
mattermost/mattermost-developer-documentation: site/content/contribute/more-info/desktop/architecture/configuration.md

---
title: "Configuration"

> heading: "Configuration and Server Management"

Reviewer: ```suggestion
heading: "Configuration and server management"
```

description: "Outlines the renderer processes that represent external Mattermost servers, and the modules that manage them."
date: 2023-04-03T00:00:00-05:00
weight: 1
aliases:
  - /contribute/desktop/architecture/configuration
---

> ### Config Module

Reviewer: ```suggestion
### Config module
```

> The **configuration** module in the **Common** module is responsible for facilitating reading from and writing to external configuration sources. It also consolidates, verifies and upgrades configuration where applicable.

Reviewer: ```suggestion
The **configuration** module in the **Common** module is responsible for facilitating reading from and writing to external configuration sources. It also consolidates, verifies, and upgrades configuration where applicable.
```

Reviewer 2: @cwarnermm I thought our style guide was not to use Oxford commas? I personally prefer them though, so that change would be better for me 😛

#### Files

We have a few different configuration files in the Desktop App, but the main one is `config.json`. Most of the user's configuration from the Settings Window is stored there, as well as any user-configured servers.

> The application supports different configuration versions, and allows for them to be migrating to the version supported by the configuration module via the `upgradePreferences` module. When no configuration is found, the `defaultPreferences` object is copied over to the main configuration module.

Reviewer 2: ```suggestion
The application supports different configuration versions and allows for them to be migrating to the version supported by the configuration module via the `upgradePreferences` module. When no configuration is found, the `defaultPreferences` object is copied over to the main configuration module.
```

> We also support a build configuration, in which the packager of the application can pre-define servers and a few other configuration items.

Reviewer 2: I had to google this one since I don't think I've ever actually written a sentence containing "in which" before even though I know I've said it out loud before
```suggestion
We also support a build configuration in which the packager of the application can pre-define servers and a few other configuration items.
```

#### Registry

> We support reading from the Windows registry to allow system administrators to define Group Policy that will pre-define servers, and potentially disable user-defined servers and automatic updates as per administrator wishes.

Reviewer 2: ```suggestion
We support reading from the Windows registry to allow system administrators to define Group Policy that will pre-define servers and potentially disable user-defined servers and automatic updates as per administrator wishes.
```

Templates for these can be found under `resources\windows\gpo`.

> ### Server Manager

Reviewer: ```suggestion
### Server manager
```

The `ServerManager` is a singleton class that acts as a single source of truth for all server configuration, managing adding/modifying/removing servers and serving the server information to the rest of the application.

#### Initialization

We populate the `ServerManager` with all servers provided by the **configuration** module, marking them as pre-defined when applicable to not allow the user to modify them. Servers are given a unique UUID when the app initialized, and this UUID acts as the global way of identifying the server to the rest of the application.

An external call is responsible for populating information about the specific Mattermost server (eg. server version, plugins installed), but the data is stored within the `ServerManager`.

#### Modification

> The `ServerManager` is the only place that allows the persistent server configuration to be modified, you cannot modify directly through the **configuration** module. Once a server is modified, the `ServerManager` will update the **configuration** module with the new changes.

Reviewer 2: ```suggestion
The `ServerManager` is the only place that allows the persistent server configuration to be modified. Changes cannot be made directly through the **configuration** module. Once a server is modified, the `ServerManager` will update the **configuration** module with the new changes.
```

When a server is added or updating, up to two events will be emitted:
- `SERVERS_UPDATE`: This event is emitted when the `ServerManager` has new changes. This could be a name, URL or an ordering change.

> - `SERVERS_URL_MODIFIED`: This event is emitted specifically when a URL has changed, signifying that the application might need to fetch new remote server information, or refresh any views associated with the server to reflect the new URL.

Reviewer 2: ```suggestion
- `SERVERS_URL_MODIFIED`: This event is emitted specifically when a URL has changed, signifying that the application might need to fetch new remote server information or refresh any views associated with the server to reflect the new URL.
```
